attr.c

```diff
@@ -744,6 +745,19 @@ static struct attr_stack *read_attr_from_index(struct index_state *istate,
 	if (!istate)
 		return NULL;
 
+	/*
+	 * In the case of cone-mode sparse-checkout, getting the
+	 * .gitattributes file from a directory is meaningless: all
+	 * contained paths will be sparse if the .gitattributes is also
+	 * sparse. In the case of a sparse index, it is critical that we
+	 * don't go looking for one as it will expand the index.
+	 */
+	init_sparse_checkout_patterns(istate);
+	if (istate->sparse_checkout_patterns &&
+	    istate->sparse_checkout_patterns->use_cone_patterns &&
+	    path_in_sparse_checkout(path, istate) == NOT_MATCHED)
```

Comment:
Maybe there would be a use in introducing a `path_in_sparse_checkout_cone()` function (which would require slight refactoring of `path_in_sparse_checkout()`)?

Reply:
This one can be simplified, I think. Let me investigate.

Reply:
I think I was focused on the cone mode because there was a bug in how non-cone-mode pattern matches worked, but that is now fixed. (Now I need to make sure that is fixed in the correct place and not in the series _after_ this one.)

Reply:
No, cone mode is necessary here because non-cone mode might still care about `.gitattributes`.